### PR TITLE
Fix Issue 23369 - Let more symbols be imported multiple times.

### PR DESCRIPTION
**Pull request by maxhaton**
The bug complains about the error message but the underlying problem here is that the error doesn't trigger at all for most other kinds of symbols.

### CODE REVIEW

File: compiler/src/dmd/declaration.d [hunk 1]
```diff
@@ -715,6 +715,7 @@ extern (C++) final class TupleDeclaration : Declaration
  */
 extern (C++) final class AliasDeclaration : Declaration
 {
+
```
Comment (ibuclaw) [inline compiler/src/dmd/declaration.d:718]:
```suggestion
```

File: compiler/src/dmd/declaration.d [hunk 2]
```diff
@@ -779,8 +780,17 @@ extern (C++) final class AliasDeclaration : Declaration
...
              * is not overloadable.
              */
-            if (type)
-                return false;
+            if (type) {
+                /*
+                    If type has been resolved already we could
+                    still be inserting an alias from an import.
+
+                    If we are handling then an alias then pretend
```
Comment (ibuclaw) [inline compiler/src/dmd/declaration.d:788]: One too many thens.

File: compiler/test/compilable/imports/aliasdump.d [hunk 3]
```diff
@@ -0,0 +1,3 @@
+alias twinkleTwinkleKillerKane = ulong;  ///
```
Comment (ibuclaw) [inline compiler/test/compilable/imports/aliasdump.d:1]: You can think of better names than that. Even `a23369` as a reference to the issue is fine.

File: compiler/src/dmd/declaration.d (hunk 1 above)
Comment (maxhaton) [reply compiler/src/dmd/declaration.d]: oops

File: compiler/src/dmd/declaration.d (hunk 2 above)
Comment (maxhaton) [reply compiler/src/dmd/declaration.d]: !

File: compiler/src/dmd/declaration.d [hunk 4]
```diff
@@ -779,8 +779,17 @@ extern (C++) final class AliasDeclaration : Declaration
             /* Semantic analysis is already finished, and the aliased entity
              * is not overloadable.
              */
-            if (type)
-                return false;
+            if (type) {
```
Comment (ibuclaw) [inline compiler/src/dmd/declaration.d:782]: Allman style (shouldn't this be picked up by the linter?)

File: compiler/test/compilable/imports/aliasdump.d [hunk 5]
```diff
@@ -0,0 +1,3 @@
+alias iainBuclaw = ulong;  ///
```
Comment (ibuclaw) [inline compiler/test/compilable/imports/aliasdump.d:1]: Very funny.

**Comment by RazvanN7** [comment]
Maybe it's just me, but I would rather get an error message in these situations. As I understood, even the person that reported the bug was actually complaining about the error message, not about the fact that the code does not compile.

File: compiler/test/compilable/importtests.d [hunk 6]
```diff
@@ -0,0 +1,5 @@
+// https://issues.dlang.org/show_bug.cgi?id=23369
+import imports.aliasdump : iainBuclaw;
+import imports.aliasdump : iainBuclaw, iainBuclaw;
+
+import imports.aliasdump : aaa;
```
Comment (RazvanN7) [inline compiler/test/compilable/importtests.d:5]: This seems unrelated to this patch.

**Comment by maxhaton** [comment]
> Maybe it's just me, but I would rather get an error message in these situations. As I understood, even the person that reported the bug was actually complaining about the error message, not about the fact that the code does not compile.

Why? It works in every other case

**Comment by maxhaton** [comment]
Redundant imports that don't conflict being an error would probably break a lot of generic code

**Comment by RazvanN7** [comment]
> Why? It works in every other case

Duplicate imports are code smells.

> Redundant imports that don't conflict being an error would probably break a lot of generic code

I don't see how generic code can introduce multiple selective imports without that being a code smell. In any case, up until this point you were not allowed to have redundant selective imports and I haven't heard anyone complain about this with regards to generic code.

To wrap it up, I just don't see the usefulness of allowing such code. If you can come up with a scenario where this is justified, I am willing to accept it, but up until now I don't see a compelling case for it.

**Comment by maxhaton** [comment]
> > Why? It works in every other case
> 
> Duplicate imports are code smells.
> 
> > Redundant imports that don't conflict being an error would probably break a lot of generic code
> 
> I don't see how generic code can introduce multiple selective imports without that being a code smell. In any case, up until this point you were not allowed to have redundant selective imports and I haven't heard anyone complain about this with regards to generic code.
> 
> To wrap it up, I just don't see the usefulness of allowing such code. If you can come up with a scenario where this is justified, I am willing to accept it, but up until now I don't see a compelling case for it.

Try it, they're allowed if the symbol can be considered an overload.

File: compiler/test/compilable/importtests.d (hunk 6 above)
Comment (maxhaton) [reply compiler/test/compilable/importtests.d:5]: It was to make sure there was no hiccups with the other forms

File: compiler/src/dmd/declaration.d (hunk 4 above)
Comment (dkorpel) [reply compiler/src/dmd/declaration.d:782]: @maxhaton ping

Comment (maxhaton) [reply compiler/src/dmd/declaration.d]: done

**Review by dkorpel — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@maxhaton I appreciate the joke, but can you please add a normal name for the alias used in the test? That is, if @ibuclaw is fine we can merge it as is.

**Comment by RazvanN7** [comment]
@dkorpel @maxhaton I didn't realize this only complains for duplicate imports of aliases.
